### PR TITLE
Improve ConfigEvalEngine performance by remove unnecessarry calls

### PR DESCRIPTION
**Pull request by yoyama**
This PR improve performance of operator processing by reducing call of Javascript engine processing.
Current code always call invokeTemplate().
But if there is no evaluated code in the text, we don't need to call it.
This PR checks text contains '$' char and only call invokeTemplate() if exists.

I tested with the following dig and measured the time with output _digdag_ attempt <id>_ .

|   secs  | 1  | 2  | 3  | Avg.  |
|:-:      |:-: |:-: |:-: |:-:    |
|original |17  |12  |13  | 14.0  |
|modified |14  | 9  | 9  | 10.67 |

As above, the PR is faster than original in this dig.
Performance depends on the dig.
But in most case it improve the performance.

- This PR improve the performance of operator by skip invokeTemplate() if needless
- Will improve CPU utilization
- Even though we move to GraalJS, this code(or way) is still effective.

```
+task1:
  for_each>:
    fruit: [apple, orange, banana]
    color: [white, black, blue, green, red, yellow]
    verb: [eat, throw, look]
  _parallel: true
  _do:
    echo>: ${verb} ${color} ${fruit}

+task2:
  for_each>:
    fruit: [apple, orange, banana]
    color: [white, black, blue, green, red, yellow]
    verb: [eat, throw, look]
  _parallel: true
  _do:
    echo>: aaaaa
```

As additional test, I measured with +task1 only WF. In this case, task always have '$' the performance should be near equivalent.

|   secs  | 1  | 2  | 3  | Avg.  |
|:-:      |:-: |:-: |:-: |:-:    |
|original |10  |7  | 6 | 7.67  |
|modified |11  |7  | 5  |7.67  |

The result is as expected. In not good condition for this PR,  performance is almost all same.

### CODE REVIEW

File: digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java [hunk 1]
```diff
@@ -213,12 +234,25 @@ private JsonNode evalValue(ObjectNode local, String code)
         }
     }
 
+    private static final Pattern requireInvokdetemplatePattern = Pattern.compile("(?m)^.*[$].*$", Pattern.MULTILINE);
```
[the remark this answers is not in the document]
Comment (frsyuki) [reply digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java]: `requireInvokdetemplatePattern.contains("$")` is even simpler and safer. but using expression seems also fair idea. which do you prefer?

File: digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java [hunk 2]
```diff
@@ -109,7 +113,8 @@ private String invokeTemplate(Invocable templateInvocable, String code, Config p
             throw new TemplateException("Failed to serialize parameters to JSON", ex);
         }
         try {
-            return (String) templateInvocable.invokeFunction("template", code, context);
+            String evaluated = (String) templateInvocable.invokeFunction("template", code, context);
+            return evaluated;
```
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java]: If there is a significant reason why this change is needed, please add comments here. If it's just styling matter, that's fine.

File: digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java [hunk 3]
```diff
@@ -213,12 +234,25 @@ private JsonNode evalValue(ObjectNode local, String code)
         }
     }
 
+    private static final Pattern requireInvokdetemplatePattern = Pattern.compile("(?m)^.*[$].*$", Pattern.MULTILINE);
+
+    @VisibleForTesting
+    protected boolean requireInvokdetemplate(String code) {
```
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java]:
```suggestion
    protected boolean requireInvokdeTemplate(String code)
    {
```

File: digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java [hunk 4]
```diff
@@ -63,6 +63,7 @@ private static String readResource(String resourceName)
 
     private final ObjectMapper jsonMapper;
     private final NashornScriptEngineFactory jsEngineFactory;
+    private DigdagMetrics metrics;
```
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java]: this variable is not used. please delete it as well as `import` lines

File: digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java]: I prefer your simple idea. Thanks!

File: digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java [hunk 5]
```diff
@@ -132,13 +137,23 @@ private String invokeTemplate(Invocable templateInvocable, String code, Config p
...
         {
             this.params = params;
-            this.templateInvocable = newTemplateInvocable(params);
+        }
+
+        // ToDo Lombok Getter(lazy=true)
+        private Invocable lazyGetTemplateInvocable()
+        {
+            if (templateInvocable == null) {
+                synchronized(this) {
```
[the remark this answers is not in the document]
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java]: Thank you for your advice. I will fix it.

File: digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java [hunk 6]
```diff
@@ -132,13 +137,23 @@ private String invokeTemplate(Invocable templateInvocable, String code, Config p
...
+        private Invocable templateInvocable = null;
         private final ImmutableList<String> noEvaluatedKeys = ImmutableList.of("_do",  "_else_do");
 
         public Context(Config params)
         {
             this.params = params;
-            this.templateInvocable = newTemplateInvocable(params);
+        }
+
+        // ToDo Lombok Getter(lazy=true)
```
[the remark this answers is not in the document]
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java]: I removed it.

File: digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java (hunk 2 above)
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java]: Sorry, I reverted it.

File: digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java (hunk 4 above)
Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java]: Thanks. Removed it.

File: digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java [hunk 7]
```diff
@@ -203,7 +212,13 @@ private JsonNode evalValue(ObjectNode local, String code)
             for (Map.Entry<String, JsonNode> pair : ImmutableList.copyOf(local.fields())) {
                 scopedParams.set(pair.getKey(), pair.getValue());
             }
-            String resultText = invokeTemplate(templateInvocable, code, scopedParams);
+            String resultText = null;
+            if (requireInvokedTemplate(code)) {
```
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java]: This is minor but you may reconsider this method name.
boolean methods used in digdag code are https://github.com/treasure-data/digdag/search?q=%22public+boolean%22&unscoped_q=%22public+boolean%22

Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java]: Thank you for your advice. I agree and  I will fix it.

Comment (yoyama) [reply digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java]: Fixed it.
